File: studentassistant/src/main/java/ru/erdenian/studentassistant/schedule/Lesson.java
package ru.erdenian.studentassistant.schedule;

import com.google.common.collect.ComparisonChain;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSortedSet;

import org.joda.time.LocalDate;
import org.joda.time.LocalTime;

import lombok.NonNull;
import lombok.Value;

/**
 * Класс пары. Является неизменяемым.
 *
 * @author Ilya Solovyev
 * @version 0.0.0
 * @since 0.0.0
 */
@Value
public final class Lesson implements Comparable<Lesson> {

    /**
     * Идентификатор пары. Должен быть уникальным для каждой пары. При редактировании существующей
     * пары должен создаваться новый объект с таким же id, старый объект должен быть удален.
     *
     * @since 0.0.0
     */
    private final long id;

    /**
     * Название предмета.
     *
     * @since 0.0.0
     */
    @NonNull
    private final String name;

    /**
     * Тип пары.
     *
     * @since 0.0.0
     */
    private final String type;

    /**
     * Массив имен преподавателей.
     *
     * @since 0.0.0
     */
    private final ImmutableSortedSet<String> teachers;

    /**
     * Массив аудиторий.
     *
     * @since 0.0.0
     */
    private final ImmutableSortedSet<String> classrooms;

    /**
     * Время начала.
     *
     * @since 0.0.0
     */
    @NonNull
    private final LocalTime startTime;

    /**
     * Время окончания.
     *
     * @since 0.0.0
     */
    @NonNull
    private final LocalTime endTime;

    /**
     * Типы повторений пары.
     *
     * @since 0.0.0
     */
    public enum RepeatType {

        /**
         * Пара повторяется в зависимости от дня недели и номера недели.
         *
         * @since 0.0.0
         */
        BY_WEEKDAY,

        /**
         * Пара повторяется в зависимости от даты.
         *
         * @since 0.0.0
         */
        BY_DATE
    }

    /**
     * Хранит тип повторения пары.
     *
     * @since 0.0.0
     */
    @NonNull
    private final RepeatType repeatType;

    /**
     * День недели, когда повторяется пара. Например, если пара повторяется по вторникам,
     * в weekday будет храниться 2.
     * Если в {@link Lesson#repeatType} хранится {@link Lesson.RepeatType#BY_WEEKDAY}, то в weekday
     * обязательно должно храниться значение от 1 (понедельник) до 7(воскресенье) включительно.
     *
     * @since 0.0.0
     */
    private final int weekday;

    /**
     * Номера недель, по которым повторяется пара. Например, если пара повторяется на 2 и 3 неделе
     * из 4, в weeks будет храниться false, true, true, false.
     * Если в {@link Lesson#repeatType} хранится {@link Lesson.RepeatType#BY_WEEKDAY}, то weeks
     * обязательно должен быть не null, и его размер должен быть больше 0.
     *
     * @since 0.0.0
     */
    private final ImmutableList<Boolean> weeks;

    /**
     * Даты, по которым повторяется пара.
     * Если в {@link Lesson#repeatType} хранится {@link Lesson.RepeatType#BY_WEEKDAY}, то dates
     * обязательно должен быть не null, и его размер должен быть больше 0.
     *
     * @since 0.0.0
     */
    private final ImmutableSortedSet<LocalDate> dates;

    /**
     * В качестве идентификатора пары используется {@link System#nanoTime()}.
     *
     * @param name       название предмета ({@link Lesson#name})
     * @param type       тип пары ({@link Lesson#type})
     * @param teachers   массив имен преподавателей ({@link Lesson#teachers})
     * @param classrooms массив аудиторий ({@link Lesson#classrooms})
     * @param startTime  время начала ({@link Lesson#startTime})
     * @param endTime    время окончания ({@link Lesson#endTime})
     * @param repeatType тип повторений ({@link Lesson#repeatType})
     * @param weekday    день недели, когда повторяется пара ({@link Lesson#weekday})
     * @param weeks      номера недель, по которым повторяется пара ({@link Lesson#weeks})
     * @param dates      даты, по которым повторяется пара ({@link Lesson#dates})
     * @since 0.0.0
     */
    public Lesson(@NonNull String name, String type,
                  ImmutableSortedSet<String> teachers, ImmutableSortedSet<String> classrooms,
                  @NonNull LocalTime startTime, @NonNull LocalTime endTime,
                  @NonNull RepeatType repeatType,
                  int weekday, ImmutableList<Boolean> weeks,
                  ImmutableSortedSet<LocalDate> dates) {

        this(System.nanoTime(), name, type, teachers, classrooms,
                startTime, endTime,
                repeatType,
                weekday, weeks,
                dates);
    }

    /**
     * @param id         идентификатор пары ({@link Lesson#id})
     * @param name       название предмета ({@link Lesson#name})
     * @param type       тип пары ({@link Lesson#type})
     * @param teachers   массив имен преподавателей ({@link Lesson#teachers})
     * @param classrooms массив аудиторий ({@link Lesson#classrooms})
     * @param startTime  время начала ({@link Lesson#startTime})
     * @param endTime    время окончания ({@link Lesson#endTime})
     * @param repeatType тип повторений ({@link Lesson#repeatType})
     * @param weekday    день недели, когда повторяется пара ({@link Lesson#weekday})
     * @param weeks      номера недель, по которым повторяется пара ({@link Lesson#weeks})
     * @param dates      даты, по которым повторяется пара ({@link Lesson#dates})
     * @since 0.0.0
     */
    public Lesson(long id, @NonNull String name, String type,
                  ImmutableSortedSet<String> teachers, ImmutableSortedSet<String> classrooms,
                  @NonNull LocalTime startTime, @NonNull LocalTime endTime,
                  @NonNull RepeatType repeatType,
                  int weekday, ImmutableList<Boolean> weeks,
                  ImmutableSortedSet<LocalDate> dates) {

        switch (repeatType) {
            case BY_WEEKDAY:
                if ((weekday < 1) || (weekday > 7)) {
                    throw new IllegalArgumentException("Некорректный номер недели: " + weekday);
                } else if (weeks == null) {
                    throw new NullPointerException("Номера недель не заданы");
                } else if (weeks.size() == 0) {
                    throw new IllegalArgumentException("Массив с номерами недель пуст");
                }
                dates = null;
                break;
            case BY_DATE:
                if (dates == null) {
                    throw new NullPointerException("Даты не заданы");
                } else if (dates.size() == 0) {
                    throw new IllegalArgumentException("Массив с датами пуст");
                }
                weekday = -1;
                weeks = null;
                break;
        }

        this.name = name;
        this.type = type;

        this.teachers = teachers;
        this.classrooms = classrooms;

        this.startTime = startTime;
        this.endTime = endTime;

        this.repeatType = repeatType;

        this.weekday = weekday;
        this.weeks = weeks;

        this.dates = dates;

        this.id = id;
    }

    @Override
    public int compareTo(@android.support.annotation.NonNull Lesson lesson) {
        return ComparisonChain.start()
                .compare(startTime, lesson.startTime)
                .compare(endTime, lesson.endTime)
                .result();
    }
}

File: entity/src/main/kotlin/ru/erdenian/studentassistant/entity/Lesson.kt
package ru.erdenian.studentassistant.entity

import android.os.Parcelable
import org.joda.time.DateTimeConstants
import org.joda.time.LocalDate
import org.joda.time.LocalTime

/**
 * Класс пары (урока).
 *
 * @property subjectName название предмета
 * @property type тип пары (лекция, семинар, ...)
 * @property teachers список преподавателей, ведущих пару
 * @property classrooms список аудиторий, в которых проходит пара
 * @property startTime время начала
 * @property endTime время конца
 * @property lessonRepeat когда повторяется пара
 * @property id уникальный id пары
 * @throws IllegalArgumentException если [subjectName] пусто или [startTime] >= [endTime]
 * @author Ilya Solovyov
 * @since 0.0.0
 */
interface Lesson : Comparable<Lesson>, Parcelable {

    val subjectName: String
    val type: String
    val teachers: ImmutableSortedSet<String>
    val classrooms: ImmutableSortedSet<String>

    val startTime: LocalTime
    val endTime: LocalTime
    val lessonRepeat: Repeat

    val semesterId: Long
    val id: Long

    override fun compareTo(other: Lesson) = compareValuesBy(
        this,
        other,
        Lesson::startTime,
        Lesson::endTime,
        Lesson::id
    )

    /**
     * Класс повторений пары.
     *
     * @author Ilya Solovyov
     * @since 0.0.0
     */
    sealed class Repeat : Parcelable {

        /**
         * Показывает, повторяется ли пара в заданный день.
         *
         * @param day день
         * @param weekNumber номер недели, содержащей этот день (начинается с 0)
         * @return true, если пара повторяется в заданный день, в противном случае false
         * @author Ilya Solovyov
         * @since 0.0.0
         */
        abstract fun repeatsOnDay(day: LocalDate, weekNumber: Int): Boolean

        /**
         * Повторение по дням недели.
         *
         * Для хранения списка недель используется List, так как массивы изменяемы.
         *
         * @property weekday день недели (1 - понедельник, 7 - воскресенье)
         * @property weeks список boolean значений, где i-е значение показывает
         *              повторять ли пару каждую i-ю неделю
         * @throws IllegalArgumentException если [weekday] задан некорректно или [weeks] пуст
         * @author Ilya Solovyov
         * @since 0.0.0
         */
        abstract class ByWeekday : Repeat() {

            abstract val weekday: Int
            abstract val weeks: List<Boolean>

            override fun repeatsOnDay(day: LocalDate, weekNumber: Int) =
                (day.dayOfWeek == weekday) && (weeks[weekNumber % weeks.size])

            /**
             * Показывает, повторяется ли пара в заданный день недели без учета номера недели.
             *
             * @param weekday день недели (1 - понедельник, 7 - воскресенье)
             * @return true, если пара повторяется в этот день недели, хотя бы на одной неделе,
             *               false в противном случае
             * @throws IllegalArgumentException если [weekday] задан некорректно
             * @author Ilya Solovyov
             * @since 0.0.0
             */
            fun repeatsOnWeekday(weekday: Int): Boolean {
                require(weekday in DateTimeConstants.MONDAY..DateTimeConstants.SUNDAY) { "Неверный день недели: $weekday" }
                return (weekday == this.weekday)
            }
        }

        /**
         * Повторение по датам.
         *
         * @property dates список дат, по которым повторяется пара
         * @throws IllegalArgumentException если [dates] пуст
         * @author Ilya Solovyov
         * @since 0.0.0
         */
        abstract class ByDates : Repeat() {

            abstract val dates: Set<LocalDate>

            override fun repeatsOnDay(day: LocalDate, weekNumber: Int) = day in dates

            /**
             * То же, что и [repeatsOnDay], но без ненужного второго параметра.
             *
             * @param date день
             * @return true, если пара повторяется в заданный день, в противном случае false
             * @author Ilya Solovyov
             * @since 0.0.0
             */
            fun repeatsOnDate(date: LocalDate) = repeatsOnDay(date, -1)
        }
    }
}
